nautobot/vpn/models.py

```diff
@@ -0,0 +1,522 @@
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "statuses",
+    "webhooks",
+)
+class VPNTunnel(PrimaryModel):  # pylint: disable=too-many-ancestors
+    """VPNTunnel model."""
+
+    name = models.CharField(max_length=CHARFIELD_MAX_LENGTH, unique=True)
+    description = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True)
+    tunnel_id = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True, verbose_name="Tunnel ID")
+    vpn_profile = models.ForeignKey(
+        to="vpn.VPNProfile",
+        on_delete=models.PROTECT,
+        related_name="vpn_tunnels",
+        blank=True,
+        null=True,
+        verbose_name="VPN Profile",
+    )
+    vpn = models.ForeignKey(
+        to="vpn.VPN",
+        on_delete=models.CASCADE,
+        related_name="vpn_tunnels",
+        blank=True,
+        null=True,
+        verbose_name="VPN",
+        help_text="VPN to which this tunnel belongs",
+    )
+    status = StatusField(blank=False, null=False)
+    role = RoleField(blank=True, null=True)
+    encapsulation = models.CharField(
+        max_length=CHARFIELD_MAX_LENGTH,
+        choices=choices.EncapsulationChoices,
+        blank=True,
+    )
+    endpoint_a = models.ForeignKey(
+        to="vpn.VPNTunnelEndpoint",
+        on_delete=models.SET_NULL,
+        related_name="endpoint_a_vpn_tunnels",
+        blank=True,
+        null=True,
+        verbose_name="Endpoint A",
+        help_text="Tunnel termination A",
+    )
+    endpoint_z = models.ForeignKey(
+        to="vpn.VPNTunnelEndpoint",
+        on_delete=models.SET_NULL,
+        related_name="endpoint_z_vpn_tunnels",
+        blank=True,
+        null=True,
+        verbose_name="Endpoint Z",
+        help_text="Tunnel termination Z",
+    )
+    tenant = models.ForeignKey(
+        to="tenancy.Tenant",
+        on_delete=models.SET_NULL,
+        related_name="vpn_tunnels",
+        blank=True,
+        null=True,
+    )
+
+    clone_fields = [
+        "description",
+        "vpn",
+        "tunnel_id",
+        "encapsulation",
+        "endpoint_a",
+        "endpoint_z",
+        "status",
+        "vpn_profile",
+        "tenant",
+        "role",
+    ]
+
+    class Meta:
+        """Meta class for VPNTunnel."""
+
+        ordering = ("name",)
+        verbose_name = "VPN Tunnel"
+
+    def __str__(self):
+        """Stringify instance."""
+        return self.name
+
+    def clean(self):
+        if self.endpoint_a and self.endpoint_z and self.endpoint_a == self.endpoint_z:
+            raise ValidationError("Endpoint A and Endpoint Z cannot be the same.")
+        return super().clean()
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class VPNTunnelEndpoint(PrimaryModel):  # pylint: disable=too-many-ancestors
+    """VPNTunnelEndpoint model."""
+
+    name = models.CharField(max_length=CHARFIELD_MAX_LENGTH, editable=False)
```

Comment:
Is it intentional that the `name` is not `unique=True` here, and additionally that there are no other `unique=True` fields or `unique_together` definition on this model?

Reply:
We may have endpoints pointing to an FQDN and this is the fallback for dynamic `name`. I can imagine that we may want the same FQDN, but with different vpn profiles, so for name it won't be unique.
But for sure there should be some `unique_together`. I'll check that with Nikos tomorrow.